File: 794-swim-in-rising-water/swim-in-rising-water.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        """
        do a BFS from initial point [(i,j), min_time] # coords, min_time to unlock
        """
        n, m = len(grid), len(grid[0])
        q = [(grid[0][0], (0, 0))]
        times = [[float('inf')] * m for _ in range(n)]
        times[0][0] = grid[0][0]

        while q:
            curr_time, curr_node = heapq.heappop(q)
            if grid[curr_node[0]][curr_node[1]] == -1: continue

            if curr_node[0] == n-1 and curr_node[1] == m-1:
                return curr_time

            grid[curr_node[0]][curr_node[1]] = -1 # mark as visited
            for x_offset, y_offset in [[-1,0], [1, 0], [0, -1], [0, 1]]:
                new_x = curr_node[0] + x_offset
                new_y = curr_node[1] + y_offset

                if 0 <= new_x < n and 0 <= new_y < m and grid[new_x][new_y] != -1:
                    new_time = max(grid[new_x][new_y], curr_time)
                    if new_time < times[new_x][new_y]:
                        times[new_x][new_y] = new_time
                        heapq.heappush(q,(new_time, (new_x, new_y)))
        
        return 0

File: 794-swim-in-rising-water/swim-in-rising-water.go
import (
    "container/heap"
)

type MinHeap [][]int // time, row, col

func (h MinHeap) Len() int {return len(h)}
func (h MinHeap) Less(i, j int) bool {return h[i][0] < h[j][0]}
func (h MinHeap) Swap(i, j int) {h[i], h[j] = h[j], h[i]}
func (h *MinHeap) Push(x any) {
    *h = append(*h, x.([]int))
}
func (h *MinHeap) Pop() any {
    old := *h
    x := old[len(old) - 1]
    *h = old[:len(old) - 1]
    return x
}

func inBounds(row, col, n int) bool {
    return row >= 0 && row < n && col >= 0 && col < n
}

func swimInWater(grid [][]int) int {
    n := len(grid)
    times := make([][]int, n)
    for i := range n {
        times[i] = make([]int, n)
        for j := range n{
            times[i][j] = (1 << 31) -1
        }
    }
    times[0][0] = grid[0][0]
    h := make(MinHeap, 0)
    heap.Init(&h)
    heap.Push(&h, []int{times[0][0], 0, 0}) // time, row, col

    for len(h) > 0 {
        currPoint := heap.Pop(&h).([]int)
        if grid[currPoint[1]][currPoint[2]] == -1 { // already visited
            continue
        }
        if currPoint[1] == n - 1 && currPoint[2] == n - 1{
            break
        }

        grid[currPoint[1]][currPoint[2]] = -1 // mark as visited

        for _, offset := range [][]int{{0, -1}, {0, 1}, {1, 0}, {-1, 0}} {
            newRow := currPoint[1] + offset[0]
            newCol := currPoint[2] + offset[1]
            if inBounds(newRow, newCol, n) && grid[newRow][newCol] != -1 {
                newTime := max(grid[newRow][newCol], currPoint[0]) 
                // can reach when path opens i.e both are submerged (max of the two)
                if newTime < times[newRow][newCol] {
                    times[newRow][newCol] = newTime
                    heap.Push(&h, []int{newTime, newRow, newCol})
                }                
            }
        }

    }
    
    return times[n-1][n-1]

}